### PR TITLE
Objdetect: update detect QRCode algorithm

### PR DESCRIPTION
**Pull request by allnes**
fix kmeans clusterization and improvement of quality
Along with this algorithm there are related data in opencv_extra:
PR link: [https://github.com/opencv/opencv_extra/pull/490](https://github.com/opencv/opencv_extra/pull/490)

### CODE REVIEW

**Comment by allnes** [comment]
@vpisarev fix done after tests with camera

File: modules/objdetect/src/qrcode.cpp [hunk 1]
```diff
@@ -704,19 +667,28 @@ double QRDecode::getQuadrilateralArea(Point2f a, Point2f b, Point2f c, Point2f d
...
-    double length_sides[3], perimeter = 0.0, triangle_area = 1.0;
-    length_sides[0] = norm(a - b);
-    length_sides[1] = norm(b - c);
-    length_sides[2] = norm(c - a);
-    for (size_t i = 0; i < 3; i++) { perimeter += length_sides[i]; }
-    perimeter /= 2;
-    for (size_t i = 0; i < 3; i++)
+    double half_perimeter = (norm(a - b) + norm(b - c) + norm(c - a)) / 2.0;
+    double triangle_area = sqrt(half_perimeter *
+                               (half_perimeter - norm(a - b)) *
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp]: `norm(a - b)` is calculated twice.

Comment (allnes) [reply modules/objdetect/src/qrcode.cpp]: @alalek done

File: modules/objdetect/src/qrcode.cpp [hunk 2]
```diff
@@ -704,19 +667,29 @@ double QRDecode::getQuadrilateralArea(Point2f a, Point2f b, Point2f c, Point2f d
...
 double QRDecode::getTriangleArea(Point2f a, Point2f b, Point2f c)
 {
-    double length_sides[3], perimeter = 0.0, triangle_area = 1.0;
-    length_sides[0] = norm(a - b);
-    length_sides[1] = norm(b - c);
-    length_sides[2] = norm(c - a);
-    for (size_t i = 0; i < 3; i++) { perimeter += length_sides[i]; }
-    perimeter /= 2;
-    for (size_t i = 0; i < 3; i++)
+    double norm_sides[] = { norm(a - b), norm(b - c), norm(c - a) };
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp]: Avoid using of arrays where they are not required.

```
double norm_c = norm(a - b),
       norm_a = norm(b - c),
       norm_b = norm(c - a);
```

File: modules/objdetect/src/qrcode.cpp [hunk 3]
```diff
@@ -704,19 +667,29 @@ double QRDecode::getQuadrilateralArea(Point2f a, Point2f b, Point2f c, Point2f d
...
+    double half_perimeter = (norm_sides[0] + norm_sides[1] + norm_sides[2]) / 2.0;
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp]:
```.diff
- (...) / 2.0
+ (...) * 0.5
```

File: modules/objdetect/src/qrcode.cpp [hunk 4]
```diff
@@ -704,19 +667,29 @@ double QRDecode::getQuadrilateralArea(Point2f a, Point2f b, Point2f c, Point2f d
...
+    double triangle_area = sqrt(half_perimeter *
+                               (half_perimeter - norm_sides[0]) *
+                               (half_perimeter - norm_sides[1]) *
+                               (half_perimeter - norm_sides[2]));
+    return triangle_area;
+}
+
+double QRDecode::getPolygonArea(vector<Point2f> points)
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp]: Why we can't use `cv::contourArea()` here instead of this method?

**Comment by alalek** [comment]
During discussion of previous PR you say that corners order are fixed.
Why we need to change order again in [testdata](https://github.com/opencv/opencv_extra/pull/490)?

Corner order must be properly defined in documentation.

File: modules/objdetect/src/qrcode.cpp [hunk 5]
```diff
@@ -35,13 +35,13 @@ class QRDecode
     Point2f intersectionLines(Point2f a1, Point2f a2, Point2f b1, Point2f b2);
     vector<Point2f> getQuadrilateral(vector<Point2f> angle_list);
     bool testBypassRoute(vector<Point2f> hull, int start, int finish);
-    double getQuadrilateralArea(Point2f a, Point2f b, Point2f c, Point2f d);
     double getTriangleArea(Point2f a, Point2f b, Point2f c);
+    double getPolygonArea(vector<Point2f> points);
     double getCosVectors(Point2f a, Point2f b, Point2f c);
```
Comment (alalek) [inline modules/objdetect/src/qrcode.cpp]: Use "static" for helper functions.

**Pull request merged**